### ISSUE TITLE
Introduce menu caching

### ISSUE THREAD

**Issue by asmecher** [pkp/pkp-lib#2933]
I reverted the late addition of menu caching as it was breaking the tests (apparently it inadvertently cached e.g. the user's logged-in vs. logged-out status, i.e. logging out would not cause the menus to list the menu accordingly).

This may be part of a bigger piece of work to replace the file-based caching, but that's not a foregone conclusion.

**Comment by asmecher** [comment]
The revert commits are listed above (once for `master` and once for `ojs-stable-3_1_0` -- the commits are identical).

**Comment by defstat** [comment]
@asmecher 
PR:
pkp-lib: #2951

**Comment by asmecher** [comment]
@defstat, did this pass the tests? (You'd need to create a PR with a submodule update as per [the wiki](https://pkp.sfu.ca/wiki/index.php?title=Github_Documentation_for_PKP_Contributors#PKP_library_submodule_changes) to get Travis to test changes that would otherwise only be in `pkp-lib`.)

**Comment by asmecher** [comment]
@defstat, it looks like the tests are failing with PHP7.1 on both OMP and OJS. The logs show:

````
[02-Nov-2017 16:18:31 UTC] PHP Fatal error:  Uncaught Error: Call to a member function getMenuId() on null in /home/travis/build/pkp/omp/lib/pkp/classes/navigationMenu/NavigationMenuItemAssignmentDAO.inc.php:304
Stack trace:
#0 /home/travis/build/pkp/omp/lib/pkp/classes/navigationMenu/NavigationMenuItemAssignmentDAO.inc.php(217): NavigationMenuItemAssignmentDAO->unCacheRelatedNavigationMenus(0)
#1 /home/travis/build/pkp/omp/lib/pkp/classes/navigationMenu/NavigationMenuItemDAO.inc.php(376): NavigationMenuItemAssignmentDAO->insertObject(Object(NavigationMenuItemAssignment))
#2 /home/travis/build/pkp/omp/lib/pkp/classes/navigationMenu/NavigationMenuDAO.inc.php(302): NavigationMenuItemDAO->installNodeSettings(0, Object(XMLNode), 1, NULL, 0, true)
#3 /home/travis/build/pkp/omp/lib/pkp/classes/install/Installer.inc.php(770): NavigationMenuDAO->installSettings(0, 'registry/naviga...')
#4 /home/travis/build/pkp/omp/lib/pkp/classes/install/Installer.inc.php(415): Installer->installDefaultNavigationMenus(Object(Install), Array)
#5 /home/travis/build/pkp/omp/lib/p in /home/travis/build/pkp/omp/lib/pkp/classes/navigationMenu/NavigationMenuItemAssignmentDAO.inc.php on line 304
[02-Nov-2017 16:19:21 UTC] PHP Warning:  assert(): assert(count($currentVersion)) failed in /home/travis/build/pkp/omp/lib/pkp/classes/core/PKPApplication.inc.php on line 342
[02-Nov-2017 16:19:21 UTC] PHP Stack trace:
[02-Nov-2017 16:19:21 UTC] PHP   1. {main}() /home/travis/build/pkp/omp/index.php:0
[02-Nov-2017 16:19:21 UTC] PHP   2. Application->execute() /home/travis/build/pkp/omp/index.php:64
[02-Nov-2017 16:19:21 UTC] PHP   3. Dispatcher->dispatch() /home/travis/build/pkp/omp/lib/pkp/classes/core/PKPApplication.inc.php:243
[02-Nov-2017 16:19:21 UTC] PHP   4. PageRouter->route() /home/travis/build/pkp/omp/lib/pkp/classes/core/Dispatcher.inc.php:134
[02-Nov-2017 16:19:21 UTC] PHP   5. PageRouter->_authorizeInitializeAndCallRequest() /home/travis/build/pkp/omp/lib/pkp/classes/core/PKPPageRouter.inc.php:233
[02-Nov-2017 16:19:21 UTC] PHP   6. IndexHandler->index() /home/travis/build/pkp/omp/lib/pkp/classes/core/PKPRouter.inc.php:372
[02-Nov-2017 16:19:21 UTC] PHP   7. IndexHandler->setupTemplate() /home/travis/build/pkp/omp/pages/index/IndexHandler.inc.php:51
[02-Nov-2017 16:19:21 UTC] PHP   8. PKPTemplateManager::getManager() /home/travis/build/pkp/omp/lib/pkp/classes/handler/PKPHandler.inc.php:439
[02-Nov-2017 16:19:21 UTC] PHP   9. TemplateManager->initialize() /home/travis/build/pkp/omp/lib/pkp/classes/template/PKPTemplateManager.inc.php:901
[02-Nov-2017 16:19:21 UTC] PHP  10. Application->getCurrentVersion() /home/travis/build/pkp/omp/lib/pkp/classes/template/PKPTemplateManager.inc.php:259
[02-Nov-2017 16:19:21 UTC] PHP  11. assert() /home/travis/build/pkp/omp/lib/pkp/classes/core/PKPApplication.inc.php:342
[02-Nov-2017 16:19:21 UTC] PHP Fatal error:  Uncaught Error: Call to a member function getVersionString() on null in /home/travis/build/pkp/omp/lib/pkp/classes/template/PKPTemplateManager.inc.php:259
Stack trace:
#0 /home/travis/build/pkp/omp/lib/pkp/classes/template/PKPTemplateManager.inc.php(901): PKPTemplateManager->initialize()
#1 /home/travis/build/pkp/omp/lib/pkp/classes/handler/PKPHandler.inc.php(439): PKPTemplateManager::getManager(Object(Request))
#2 /home/travis/build/pkp/omp/pages/index/IndexHandler.inc.php(51): PKPHandler->setupTemplate(Object(Request))
#3 /home/travis/build/pkp/omp/lib/pkp/classes/core/PKPRouter.inc.php(372): IndexHandler->index(Array, Object(Request))
#4 /home/travis/build/pkp/omp/lib/pkp/classes/core/PKPPageRouter.inc.php(233): PKPRouter->_authorizeInitializeAndCallRequest(Array, Object(Request), Array, false)
#5 /home/travis/build/pkp/omp/lib/pkp/classes/core/Dispatcher.inc.php(134): PKPPageRouter->route(Object(Request))
#6 /home/travis/build/pkp/omp/lib/pkp/classes/core/PKPApplication.inc.php(243): Dispatcher->dispatch( in /home/travis/build/pkp/omp/lib/pkp/classes/template/PKPTemplateManager.inc.php on line 259
````

**Comment by asmecher** [comment]
@defstat -- thanks, it looks like the tests are passing now. Just a couple of small comments, then please go ahead and commit/port to stable. Anecdotally, I think I can see an improvement on the time the Travis tests are taking.

**Comment by asmecher** [comment]
Thanks, @defstat! Merged, but *not* back-ported to `ojs-stable-3_1_0` (at least until we've had it deployed on a few more dev boxes).

**Comment by defstat** [comment]
@asmecher
PRs ojs-stable-3_1_0:
pkp-lib: #3024
OJS: pkp/ojs#1669

**Comment by asmecher** [comment]
(Leaving this unmerged for a little while as we assess its stability.)

**Comment by asmecher** [comment]
@defstat, now that OJS 3.1.0-1 is out, I think it's a good time to merge these into the `master` branch (for OJS and OMP). I suggest we don't merge into `ojs-stable-3-1_0`, at least until it's had some time to ripen in `master`.

**Comment by defstat** [comment]
@asmecher I think caching has already been merged to master branch

**Comment by asmecher** [comment]
Confirmed, thanks!
